src/lib/ppx_coda/ppx_representatives/ppx_representatives.ml

```diff
@@ -0,0 +1,273 @@
+open Base
+open Ppxlib
+
+let deriver_name = "to_representatives"
+
+let mangle ~suffix name =
+  if String.equal name "t" then suffix else name ^ "_" ^ suffix
+
+let mangle_lid ~suffix lid =
+  match lid with
+  | Lident name ->
+      Lident (mangle ~suffix name)
+  | Ldot (lid, name) ->
+      Ldot (lid, mangle ~suffix name)
+  | Lapply _ ->
+      assert false
+
+let mk_lid name = Loc.make ~loc:name.loc (Longident.Lident name.txt)
+
+let constr_of_decl ~loc decl =
+  Ast_builder.Default.ptyp_constr ~loc (mk_lid decl.ptype_name)
+    (List.map ~f:fst decl.ptype_params)
+
+let is_builtin = function
+  | "unit" | "bool" | "int" | "string" | "char" ->
+      true
+  | _ ->
+      false
+
+(* The way that we expand representatives below makes it extremely easy to blow
+   the stack if we're not tail-recursive. All generated list iterators should
+   use the [rev_*] versions of functions when their non-reversed alternatives
+   are not tail recursive.
+*)
+let rec core_type ~loc (typ : core_type) : expression =
+  let open Ast_builder.Default in
+  match typ.ptyp_desc with
+  | Ptyp_any ->
+      [%expr Stdlib.Lazy.from_fun (fun () -> failwith "Unknown type")]
+  | Ptyp_var name ->
+      (* Names for type variables should be placed in context at the type
+        declaration level.
+     *)
+      evar ~loc name
+  | Ptyp_arrow (label, _, _) ->
```

Comment:
For this type:
```ocaml
type t = int -> string
```
the generated code is:
```ocaml
  }]
type t = int -> string[@@deriving to_representatives]
let (to_representatives : t list lazy_t) =
  lazy (fun _ -> failwith "Functional type")
```
which gives a compile error:
```
Error: This expression should not be a function, the expected type is
t list
```
Would it be better to raise an error statically?

Reply:
This was a typo, should be fixed now. I don't imagine that we will end up using this for functions, but we can make a value of the right shape, so it seemed better to me to return that instead of having an error and potentially blocking useful output.